Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1338 @@
+* `EUnit`: The unit literal, with type `()`. `()`.
+* `EList`: A list literal. `[1, my_var, if true then 1 else 0`].
+* `EPar`: An expression within parentheses. `(1 + foo bar)`.
+* `EApp`: The application of an expression in a function. `my_func 1 "foo"`.
+* `EIfThen`: A conditional producing a unit. `if true then print "Hello"`.
+* `EIfThenElse`: A conditional. `if true then 1 else 2`.
+* `ENeg`: The negation of an expression. `-(x + 5)`.
+* `EBinOp`: The application of a binary operator to two expressions. `2 * 3 - 5`.
+* `EOp`: A binary operator without operands. `(+) x 1`.
+* `ELetIn`: Locally declares a variable. `let x = 5 in my_function x`.
+
+We've previously defined a few primitives, namely `integer`, `name`, and `string`.
+We may now add a few more simple productions.
+Just so we're on the same page, I'll repeat the productions of `exp` that we already had.
+
+```happy
+exp :: { Exp L.Range }
+  : integer                  { unTok $1 (\range (L.Integer int) -> EInt range int) }
+  | name                     { EVar (info $1) $1 }
+  | string                   { unTok $1 (\range (L.String string) -> EString range string) }
+  | '(' ')'                  { EUnit (L.rtRange $1 <-> L.rtRange $2) }
+  | '[' sepBy(exp, ',') ']'  { EList (L.rtRange $1 <-> L.rtRange $3) $2 }
+  | '(' exp ')'              { EPar (L.rtRange $1 <-> L.rtRange $3) $2 }
+```
+
+The only thing new here is the introduction of a new `sepBy` function, which should separate various productions (`exp`) by a separator (`,`), just like a list in Haskell.
+The definition of `sepBy` is provided below.
+
+```happy
+sepBy_rev(p, sep)
+  :                     { [] }
+  | sepBy(p, sep) sep p { $3 : $1 }
+
+sepBy(p, sep)
+  : sepBy_rev(p, sep) { reverse $1 }
+```
+
+Parsing such "atoms" is simple enough, but now we will find difficulties in parsing the other types of expressions, as they will have more complicated interactions.
+
+#### Parsing function applications
+
+Let's begin with `EApp`. You might want to add a new production defining it like so:
+
+```happy
+  | exp exp                  { EApp (info $1 <-> info $2) $1 $2 }
+```
+
+But immediately see that it will produce five shift/reduce conflicts.
+The ambiguity comes from the fact that `a b c` could be parsed either as `(a b) c` or `a (b c)`.
+If you analyze the output of `happy -i src/Parser.y`, you should see that it can either shift to another token or reduce it to a function application.
+
+Remember, shifting asks Happy to continue parsing, making the output right-associative, while reducing asks Happy to accept what we have so far, making the output left-associative.
+Since shifting is the default action, we will get the incorrect output with `let x = a b c`:
+
+```haskell
+Right
+  [ Dec _ (Name _ "x") [] Nothing (EApp _ (EVar _ (Name _ "a")) (EApp _ (EVar _ (Name _ "b")) (EVar _ (Name _ "c"))))
+  ]
+```
+
+Unfortunately, Happy doesn't have a `%reduce` directive, only a `%shift` directive.
```

Comment:
Do you mean to use it like `exp exp %prec APP`? I could mention it, but in this situation, it causes a lot of shift/reduce and reduce/reduce conflicts, at least when using it with the finished grammar. I will mention it but continue with the current solution to avoid refactoring the current grammar and changing the article too much.

Reply:
>  a lot of shift/reduce and reduce/reduce conflicts

Not if you list all terminals in the precedence table. But yes, that is annoying. There is however an argument to be made that it is often less annoying than rewriting the grammar.

Reply:
I see, thanks for explaining. I added a few more words about it.